### DISCUSSION TITLE
Client Side Mouse

### DISCUSSION THREAD

**Discussion by FallingSnow** [LizardByte/community#235] [Sunshine]
I'm interested in adding a mouse mode that uses the client side mouse with the host cursor image to lower perceived latency. I'd like this mode to be toggle-able as one of the modes toggled by `Ctrl + Alt + Shift + M`.

This should be similar to the current client side mouse mode where you can see both the client cursor and host cursor. However the host cursor would be hidden and client side cursor should take on the same image as the host cursor.

Is the host side mouse composited into the image on the host side or is the cursor image/plane transferred to the client, then composited into the output?

**Comment by cgutman** [comment]
Yep this is on our radar. The cursor is encoded into the video stream by Sunshine (as it was with GeForce Experience too).

The tricky part is that there are a lot of different types of cursor images (some require XOR blending with the screen contents), so there will need to be a negotiation process between clients and hosts to determine which types of cursors (if any) are eligible for client-side drawing. Of course, it also requires updates to each of the Moonlight clients.

We shouldn't need a special key combo for it. It will just offload cursor drawing to the client any time a compatible cursor image is present.

**Reply by FallingSnow** [reply]
> some require XOR blending with the screen contents

Are you specifically referring to cursors with alpha?

> We shouldn't need a special key combo for it. It will just offload cursor drawing to the client any time a compatible cursor image is present.

Ah yes, you are right.

**Reply by cgutman** [reply]
> Are you specifically referring to cursors with alpha?

Nope, there are those too but some cursors (like the I-beam) actually invert the screen contents underneath with an XOR mask.

https://learn.microsoft.com/en-us/windows/win32/api/dxgi1_2/ne-dxgi1_2-dxgi_outdupl_pointer_shape_type
